main/lsp/LSPTask.h

```diff
@@ -53,12 +96,53 @@ class LSPRequestTask : public LSPTask {
 protected:
     const MessageId id;
 
-    LSPRequestTask(const LSPConfiguration &config, MessageId id, bool enableMultithreading = false);
+    LSPRequestTask(const LSPConfiguration &config, MessageId id, LSPMethod method);
 
     virtual std::unique_ptr<ResponseMessage> runRequest(LSPTypecheckerDelegate &typechecker) = 0;
 
 public:
     void run(LSPTypecheckerDelegate &typechecker) override;
+
+    // Requests cannot override this method, as runRequest must be run (and it only runs during the RUN phase).
+    Phase finalPhase() const override;
+
+    bool cancel(const MessageId &id) override;
+};
+
+/**
+ * A special form of LSPTask that has direct access to the typechecker and controls its own scheduling.
+ * Is only used for slow path-related tasks. Do not use for anything else.
+ */
+class LSPDangerousTypecheckerTask : public LSPTask {
```

Comment:
This is the main abstraction for dangerous tasks that need direct access to the typechecker. Only a few tasks are of this shape.

These are special because I want to make `LSPTask` a golden path for new LSP features. Pretty much every typechecker feature that LSPTasks have access to are benign and won't cause subtle bugs. The same is not true of these dangerous tasks.